### PR TITLE
Allow customization of fetch patterns

### PR DESCRIPTION
**Pull request by syscafedevelopment**
Customization of fetch patterns including specifying the RequestStage and ResourceType

### CODE REVIEW

**Comment by kblok** [comment]
Hey @syscafedevelopment. Could you tell me a real use case scenario?
So I can try to propose this upstream (a.k.a puppeteer).

**Comment by kblok** [comment]
We would also need some tests for this.

And by the way. Thank you for your time contributing :)

**Comment by syscafedevelopment** [comment]
Hey @kblok, 

A scenario would be:
- the dev wants to download a certain resource (e.g., a video, a pdf, etc.)
- when the request interception is made, the dev will need to call Chrome's [Fetch.getResponseBody](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-getResponseBody) or [Fetch.takeResponseBodyAsStream](https://chromedevtools.github.io/devtools-protocol/tot/Fetch#method-takeResponseBodyAsStream)
- however, this will require the interception to happen at `HeadersReceived` stage
- additionally, the dev only needs to intercept those specific resource types, having predefined URL patterns
- at the moment, all interceptions are at `Request` stage, with a hardcoded pattern of `"*"`, which does not allow response body interceptions, nor selective interceptions

Apparently there are several complains and complicated workarounds about allowing downloading things via Puppeteer, which would start to be solved by customizing the interception stage first.

About unit tests, I'll handle those as soon as possible.

File: lib/PuppeteerSharp/Messaging/RequestWillBeSentPayload.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@ internal class RequestWillBeSentPayload
         public string LoaderId { get; set; }
         public Payload Request { get; set; }
 
-        public ResponsePayload RedirectResponse { get; set; }
```
Comment (syscafedevelopment) [inline lib/PuppeteerSharp/Messaging/RequestWillBeSentPayload.cs]: I wonder if this is ever used, considering it has never been set before

**Comment by syscafedevelopment** [comment]
@kblok, any idea why this is not passing? It seems it needs Mono.Posix assembly? Why?

**Comment by kblok** [comment]
@syscafedevelopment it seems that ShouldEmitCreatedAndDestroyedEvents is getting stuck...

On the other hand, I like this feature, but I'm waiting for Puppeteer's feedback here https://github.com/GoogleChrome/puppeteer/issues/4624. 
I don't want us to implement a feature and then they implement the same feature in a different way and we would be forced to change our API.

**Comment by kblok** [comment]
@syscafedevelopment could you share your case here https://github.com/GoogleChrome/puppeteer/issues/4624 ?

**Comment by emiliavanderwerf** [comment]
@kblok @syscafedevelopment Is this planned to be merged soon please?

**Comment by kblok** [comment]
@emiliavanderwerf I could try to push a little bit upstream to see if we get this implemented in puppeteer.

**Comment by emiliavanderwerf** [comment]
@kblok Ah, so not even [Puppeteer](https://github.com/puppeteer/puppeteer/blob/v7.0.0/docs/api.md) implements this yet?

**Comment by kblok** [comment]
@emiliavanderwerf exactly.

**Pull request closed without merging**